### PR TITLE
added-announcement

### PR DESCRIPTION
**Pull request by solomiyon**
dev
## JIRA

* [Main JIRA ticket](https://jira.softserve.academy/secure/RapidBoard.jspa?rapidView=id)

## Code reviewers

- [ ] @github_username

### Second Level Review

- [ ] @github_username

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

**Review by sharix4 — CHANGES_REQUESTED** [review]

File: EPlast/EPlast.BLL/Services/GoverningBodies/Announcement/GoverningBodyAnnouncementService.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        {
+            _repoWrapper = repositoryWrapper;
+            _mapper = mapper;
+            _context = context;
+            _userManager = userManager;
+        }
+
+        public async Task<bool> AddAnnouncement(string text)
+        {
+            if (text != null)
```
Comment (sharix4) [inline EPlast/EPlast.BLL/Services/GoverningBodies/Announcement/GoverningBodyAnnouncementService.cs:36]: if (text == null)
            {
return false; 
}

File: EPlast/EPlast.Tests/Controllers/GoverningBodiesControllerTests.cs [hunk 2]
```diff
@@ -470,6 +476,113 @@ public async Task GetUserPreviousAdministrations_Valid_Test()
...
+        public async Task GetById_Valid()
+        {
+            //Arrange
+            _governingBodyAnnouncementService.Setup(g => g.GetAnnouncementById(It.IsAny<int>()))
+                .ReturnsAsync(new GoverningBodyAnnouncementUserDTO());
+
+            //Act
+            var result = await _governingBodiesController.GetById(It.IsAny<int>());
+            var resultValue = (result as ObjectResult).Value;
+            //Assert
```
Comment (sharix4) [inline EPlast/EPlast.Tests/Controllers/GoverningBodiesControllerTests.cs:564]: spaсe

File: EPlast/EPlast.Tests/Services/GoverningBody/Announcement/GoverningBodyAnnouncementServiceTest.cs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+                .ReturnsAsync(nullGoverningBodyAnnouncement);
+
+            //Assert
+            Exception exception = Assert.ThrowsAsync(typeof(ArgumentNullException),
+               async () => { await _governingBodyAnnouncementService.DeleteAnnouncement(It.IsAny<int>()); });
+            Assert.AreEqual("Value cannot be null. (Parameter 'Announcement with 0 not found')", exception.Message);
+        }
+
+
+        //_repoWrapper.User.GetFirstOrDefaultAsync(d => d.Id == announcement.UserId
```
Comment (sharix4) [inline EPlast/EPlast.Tests/Services/GoverningBody/Announcement/GoverningBodyAnnouncementServiceTest.cs:96]: delete

**Review by sharix4 — CHANGES_REQUESTED** [review]

File: EPlast/EPlast.Tests/Controllers/GoverningBodiesControllerTests.cs [hunk 4]
```diff
@@ -488,6 +601,15 @@ private string GetStringTestId()
             GoverningBodyDocuments = new List<GoverningBodyDocumentsDTO>()
         };
 
+        //private GoverningBodyAnnouncementUserDTO CreateGoverningBodyAnnouncementUserDTO => new GoverningBodyAnnouncementUserDTO()
```
[the remark this answers is not in the document]
Comment (solomiyon) [reply EPlast/EPlast.Tests/Controllers/GoverningBodiesControllerTests.cs]: done

File: EPlast/EPlast.BLL/Services/GoverningBodies/Announcement/GoverningBodyAnnouncementService.cs (hunk 1 above)
Comment (solomiyon) [reply EPlast/EPlast.BLL/Services/GoverningBodies/Announcement/GoverningBodyAnnouncementService.cs:36]: but this works in the same way

File: EPlast/EPlast.Tests/Controllers/GoverningBodiesControllerTests.cs (hunk 2 above)
Comment (solomiyon) [reply EPlast/EPlast.Tests/Controllers/GoverningBodiesControllerTests.cs:565]: done

File: EPlast/EPlast.Tests/Services/GoverningBody/Announcement/GoverningBodyAnnouncementServiceTest.cs (hunk 3 above)
Comment (solomiyon) [reply EPlast/EPlast.Tests/Services/GoverningBody/Announcement/GoverningBodyAnnouncementServiceTest.cs]: done

File: EPlast/EPlast.BLL/Interfaces/GoverningBodies/Announcement/IGoverningBodyAnnouncementService.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace EPlast.BLL.Interfaces.GoverningBodies
+{
+    public interface IGoverningBodyAnnouncementService
+    {
+        Task<IEnumerable<GoverningBodyAnnouncementUserDTO>> GetAllAnnouncementAsync();
+        Task DeleteAnnouncement(int id);
```
[the remark this answers is not in the document]
Comment (solomiyon) [reply EPlast/EPlast.BLL/Interfaces/GoverningBodies/Announcement/IGoverningBodyAnnouncementService.cs]: done

File: EPlast/EPlast.BLL/Interfaces/GoverningBodies/Announcement/IGoverningBodyAnnouncementService.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+        Task<bool> AddAnnouncement(string text);
```
[the remark this answers is not in the document]
Comment (solomiyon) [reply EPlast/EPlast.BLL/Interfaces/GoverningBodies/Announcement/IGoverningBodyAnnouncementService.cs]: done

File: EPlast/EPlast.BLL/Interfaces/GoverningBodies/Announcement/IGoverningBodyAnnouncementService.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+        Task<GoverningBodyAnnouncementUserDTO> GetAnnouncementById(int id);
```
[the remark this answers is not in the document]
Comment (solomiyon) [reply EPlast/EPlast.BLL/Interfaces/GoverningBodies/Announcement/IGoverningBodyAnnouncementService.cs]: done

File: EPlast/EPlast.Tests/Services/GoverningBody/Announcement/GoverningBodyAnnouncementServiceTest.cs [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+            var a = _repoWrapper.Setup(u => u.User.GetFirstOrDefaultAsync(It.IsAny<Expression<Func<User, bool>>>(),
+              It.IsAny<Func<IQueryable<User>,
+              IIncludableQueryable<User, object>>>())).ReturnsAsync(new User());
+            _mapper.Setup(m => m.Map<IEnumerable<UserDTO>>(a));
+
+            //Act
+            var result = _governingBodyAnnouncementService.GetAllUserAsync();
+
+            //Assert
+            Assert.IsNotNull(result);
```
Comment (danylo-melnychuk) [inline EPlast/EPlast.Tests/Services/GoverningBody/Announcement/GoverningBodyAnnouncementServiceTest.cs:152]: Probably, we also have to Assert.Equal().
What do you think Solomiya?

File: EPlast/EPlast.WebApi/Controllers/GoverningBodiesController.cs [hunk 9]
```diff
@@ -274,5 +285,51 @@ public async Task<IActionResult> GetUserPreviousAdministrations(string UserId)
...
         }
+
+        [HttpPost("AddAnnouncement/{text}")]
+        [Authorize(AuthenticationSchemes = "Bearer")]
+        public async Task<IActionResult> AddAnnouncement(string text)
+        {
+            if (ModelState.IsValid)
+            {
+                await _governingBodyAnnouncementService.AddAnnouncement(text);
+                return NoContent();
```
Comment (danylo-melnychuk) [inline EPlast/EPlast.WebApi/Controllers/GoverningBodiesController.cs:296]: Are you sure that we have to use status code 204 either 200?

File: EPlast/EPlast.BLL/Services/GoverningBodies/Announcement/GoverningBodyAnnouncementService.cs [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+            {
+                var governingBodyAnnouncementDTO = new GoverningBodyAnnouncementDTO();
+                governingBodyAnnouncementDTO.Text = text;
+                governingBodyAnnouncementDTO.UserId = _userManager.GetUserId(_context.HttpContext.User);
+                var announcement = _mapper.Map<GoverningBodyAnnouncementDTO, GoverningBodyAnnouncement>(governingBodyAnnouncementDTO);
+                announcement.Date = DateTime.Now;
+                await _repoWrapper.GoverningBodyAnnouncement.CreateAsync(announcement);
+                await _repoWrapper.SaveAsync();
+                return true;
```
Comment (ostappppppp) [inline EPlast/EPlast.BLL/Services/GoverningBodies/Announcement/GoverningBodyAnnouncementService.cs:45]: if I`m not mistaken, we separate `return` statement with empty line in whole project
it's used for better reading experience

File: EPlast/EPlast.WebApi/Controllers/GoverningBodiesController.cs (hunk 9 above)
Comment (ostappppppp) [reply EPlast/EPlast.WebApi/Controllers/GoverningBodiesController.cs:296]: +1
204 used for deletion operations

File: EPlast/EPlast.BLL/Services/GoverningBodies/Announcement/GoverningBodyAnnouncementService.cs (hunk 10 above)
Comment (solomiyon) [reply EPlast/EPlast.BLL/Services/GoverningBodies/Announcement/GoverningBodyAnnouncementService.cs:46]: fixed

File: EPlast/EPlast.WebApi/Controllers/GoverningBodiesController.cs (hunk 9 above)
Comment (solomiyon) [reply EPlast/EPlast.WebApi/Controllers/GoverningBodiesController.cs]: fixed

File: EPlast/EPlast.Tests/Services/GoverningBody/Announcement/GoverningBodyAnnouncementServiceTest.cs (hunk 8 above)
Comment (solomiyon) [reply EPlast/EPlast.Tests/Services/GoverningBody/Announcement/GoverningBodyAnnouncementServiceTest.cs:154]: added more asserts

**Review by OleksandrBilan — APPROVED** [review]

**Review by christykobyletska — APPROVED** [review]

**Review by nicoolle — APPROVED** [review]

**Review by sharix4 — CHANGES_REQUESTED** [review]

**Review by ostappppppp — APPROVED** [review]

File: EPlast/EPlast.BLL/Services/GoverningBodies/Announcement/GoverningBodyAnnouncementService.cs (hunk 1 above)
Comment (solomiyon) [reply EPlast/EPlast.BLL/Services/GoverningBodies/Announcement/GoverningBodyAnnouncementService.cs]: fixed

**Review by sharix4 — APPROVED** [review]

**Pull request merged**